### ISSUE TITLE
ModuleNotFoundError: Module not found: Error: Can't resolve '@gatsbyjs/reach-router/lib/utils' in '/.cache'

### ISSUE THREAD

**Issue by JustFly1984** [gatsbyjs/gatsby#36565]
### Preliminary Checks

- [X] This issue is not a duplicate. Before opening a new issue, please search existing issues: https://github.com/gatsbyjs/gatsby/issues
- [X] This issue is not a question, feature request, RFC, or anything other than a bug report directly related to Gatsby. Please post those things in GitHub Discussions: https://github.com/gatsbyjs/gatsby/discussions

### Description

I'm getting issue running `yarn dev`:

```
ModuleNotFoundError: Module not found: Error: Can't resolve '@gatsbyjs/reach-router/lib/utils' in '/Users/user/projects/personal/project/.cache'
```

and running `yarn build`:

```
failed Building production JavaScript and CSS bundles - 63.753s

 ERROR #98124  WEBPACK

Generating JavaScript bundles failed

Can't resolve '@gatsbyjs/reach-router/lib/utils' in
'/Users/user/projects/personal/prject/.cache'

If you're trying to use a package make sure that '@gatsbyjs/reach-router/lib/utils' is
installed. If you're trying to use a local file make sure that the path is correct.

File: .cache/find-path.js:1:0


 ERROR #98124  WEBPACK

Generating JavaScript bundles failed

Can't resolve '@gatsbyjs/reach-router/lib/history' in
'/Users/user/projects/personal/project/.cache'

If you're trying to use a package make sure that '@gatsbyjs/reach-router/lib/history' is
installed. If you're trying to use a local file make sure that the path is correct.

File: .cache/navigation.js:1:394


error Command failed with exit code 1.
```

before that I had to force install gatsby-scripts, gatsby-react-router-scroll, gatsby-link into package.json, I have `@gatsbyjs/reach-router` installed and new Head API refactored, but can't run dev, due an error above

It is just not looking down the node_modules tree. Please fix.

crosslink to the disscussion: https://github.com/gatsbyjs/gatsby/discussions/35313

### Reproduction Link

no repro

### Steps to Reproduce

Just running plain project with "gatsby": "4.22.0"

### Expected Result

Should run `yarn dev` without errors

### Actual Result

ModuleNotFoundError: Module not found: Error: Can't resolve '@gatsbyjs/reach-router/lib/utils' in '/.cache'

### Environment

```shell
System:
    OS: macOS 12.5.1
    CPU: (16) x64 Intel(R) Core(TM) i9-9980HK CPU @ 2.40GHz
    Shell: 5.8.1 - /bin/zsh
  Binaries:
    Node: 16.16.0 - ~/.nvm/versions/node/v16.16.0/bin/node
    Yarn: 1.22.19 - /usr/local/bin/yarn
    npm: 8.11.0 - ~/.nvm/versions/node/v16.16.0/bin/npm
  Browsers:
    Chrome: 105.0.5195.102
    Firefox: 102.0.1
    Safari: 15.6.1
  npmPackages:
    gatsby: 4.22.0 => 4.22.0
    gatsby-link: 4.22.0 => 4.22.0
    gatsby-plugin-catch-links: 4.22.0 => 4.22.0
    gatsby-plugin-google-fonts: 1.0.1 => 1.0.1
    gatsby-plugin-minify-classnames: 4.1.3 => 4.1.3
    gatsby-plugin-root-import: 2.0.8 => 2.0.8
    gatsby-plugin-typescript: 4.22.0 => 4.22.0
    gatsby-react-router-scroll: 5.22.0 => 5.22.0
    gatsby-script: 1.7.0 => 1.7.0
```

### Config Flags

no

**Comment by LekoArts** [comment]
As said in https://github.com/gatsbyjs/gatsby/discussions/35313, please open only an issue if you can provide a minimal reproduction. Thanks!

**Comment by JustFly1984** [comment]
I've solved an issue. The issue was "@gatsbyjs/reach-router": "1.3.10-no-navigate.4", which has been installed by yarn as latest version. Rolling back to "@gatsbyjs/reach-router": "1.3.9", solved an issue.

For some reaason no-navigate version gets installed by default if you run `yarn add @gatsbyjs/reach-router`, instead of 1.3.9

While inspecting `yarn-lock` I see

```
gatsby@4.22.0:
  version "4.22.0"
  resolved "https://registry.yarnpkg.com/gatsby/-/gatsby-4.22.0.tgz#45abc807f209f1be1872d2c6c86f7f76b4a61c33"
  integrity sha512-tpssrpJX0vn27oolu72XaDyXpqLPsAh7EjebeOVOlegEq6ZwHg6b+XjhQoLFujt3ogXHza+vNhSuMQ6NFLzzMA==
  dependencies:
    "@babel/code-frame" "^7.14.0"
    "@babel/core" "^7.15.5"
    "@babel/eslint-parser" "^7.15.4"
    "@babel/helper-plugin-utils" "^7.14.5"
    "@babel/parser" "^7.15.5"
    "@babel/runtime" "^7.15.4"
    "@babel/traverse" "^7.15.4"
    "@babel/types" "^7.15.4"
    "@builder.io/partytown" "^0.5.2"
    "@gatsbyjs/reach-router" "^1.3.9"
```

The ^ sign allows to install 1.3.10, so here is an error.

I had to add "@gatsbyjs/reach-router": "1.3.9" into `resolutions` field in `package.json` and regenerate `yarn-lock` and `node_modules/` to fix everything

**Comment by JustFly1984** [comment]
@LekoArts After updating gatsby from 4.21 to 4.22, it started to fail building `ModuleNotFoundError: Module not found: Error: Can't resolve in .cache` first `gatsby-scripts`, after installing gatsby-scripts it fails with `gatsby-react-router-scroll`, after installing it, it fails missing `@gatsbyjs/reach-router`, after installing it, it istalls @gatsbyjs/reach-router@1.3.10-no-navigate version and fails on `@gatsbyjs/reach-router/lib/utils`, which is fixable by pinning version in resolutions.

There is something deeply wrong with transitive ^ dependencies.

**Comment by LekoArts** [comment]
We can only help with a minimal reproduction, our current CI doesn't report any issues and most other users also don't report an issue about this. So it must be something specific to your setup and thus we'll need a reproduction.

**Comment by barbalex** [comment]
I understand your point.

But from our point of view: We have several gatsby projects (and also a number of others). It has become unbearable having complex dependency issues way too often when updating dependencies. 

Any update of dependencies (also non-gatsby-related ones) can lead to similar problems like this one. Probably any change to package-lock.json or yarn.lock.

On projects building on vite, electron and nextjs this has been way better.

For this reason we are migrating away from gatsby. New projects are built in vite or nextjs. Old ones getting completely blocked by such issues are migrated to vite/nextjs. Which hurts. But getting the development process randomly blocked is not tolerable.

**Comment by JustFly1984** [comment]
@LekoArts 

@barbalex I'm with you at this point. This was last nail in coffin for create-react-app. You are pulling your hair out trying to upgrade it to the next major version. But gatsbyjs... you are pulling your hair out with any minor version update. This is unbearable. I can't in the right mind recommend gatsby as stable platform anymore. I'm using it only for existing projects, and it is a PAIN in compare with bearable next.js

New Head API did broke one of my projects build due to missing wrapRootElement for Head components, which is crucial for createPage API and react-intel or redux dependent projects. And the fix is promised in v5, which is not even on the horizon.

Please improve your package maintainability and ease of updating. You have so many features, that you are foggeting about them while implementing next features.

I hope new Layout API supports wrapRootElement at the start, cos using redux or react-intel, or apollo stuff will be not possible as it is with Head API right now.

For a long time Gatsby was a breath of use, but now it looks like slowdown. Please don't kill such great project by avoiding transitive dependencies issues. It compromises not only dev experience, but also a painpoint for security, cos allowing ^ in the version provides malicious actors an opportunity to inject malicious new version of any transitive dependency.

I would argue that you should carefully inspect all gatsby packages, and remove ^ sign from versions. Github's dependabot would help to identify outdated/unsecure dependencies, and help to provide secure software.

I'm very happy that now we can use Typescript for config files, so you are in lead on that against next.js, which can by now use only comments to improve some type inference in js files.

I'm using Gatsby for 5+ years, and hope it will release v5 sooner! Big Thanks to all maintainers and community!

**Comment by JWesorick** [comment]
I was experiencing this issue too. I'm going from v2 => v4. The issue went away once I removed the old `gatsby-image` includes and cleared the cache `rm -r .cache`.

**Comment by adamatronix** [comment]
I'm getting `ModuleNotFoundError: Module not found: Error: Can't resolve 'gatsby-script'` after updating all my gatsby stuff to the latest. 4.23.1.

**Comment by barbalex** [comment]
Same here once again with 4.24

**Comment by aaron5670** [comment]
Same problem here when we upgrade from React 17.0.2 to 18.2.
We are on Gatsby 4.22.1.

Any solution to this? Since there are several open issues on Github with the same problem...

**Comment by theoBLT** [comment]
Having the same error messages upgrading from V4 to V5.  Same sentiment, too, that I should probably migrate away. Updating each of the above mentioned package manually fixed the problem ¯\_(ツ)_/¯
